pkg/vault/contracts/authorizer/TimelockAuthorizer.sol

```diff
@@ -513,25 +514,72 @@ contract TimelockAuthorizer is IAuthorizer, IAuthentication, ReentrancyGuard {
+        require(_isDelayShorterThanSetAuthorizer(delay), "DELAY_EXCEEDS_SET_AUTHORIZER");
+
+        _grantDelays[actionId] = delay;
+        emit GrantDelaySet(actionId, delay);
+    }
+
+    function _isDelayShorterThanSetAuthorizer(uint256 delay) private view returns (bool) {
+        // No delay can be greater than the current delay for changing the Authorizer itself (`IVault.setAuthorizer`).
+        // Otherwise, it'd be possible to execute the action with a shorter delay by simply replacing the
+        // TimelockAuthorizer with a different contract that didn't enforce these delays.
+        // Note that it is still possible for an action to end up with a delay shorter than `setAuthorizer` if e.g.
+        // `setAuthorizer`'s delay was to ever be decreased, but this is not expected to happen. The following check is
+        // therefore simply a way to try to prevent user error, but is not infallible.
+
+        bytes32 setAuthorizerActionId = _vault.getActionId(IVault.setAuthorizer.selector);
+        return delay <= _delaysPerActionId[setAuthorizerActionId];
+    }
+
     /**
-     * @notice Schedules an execution to set action `actionId`'s delay to `newDelay`.
+     * @notice Schedules an execution to set the delay for `actionId`' to `newDelay`.
      */
     function scheduleDelayChange(
         bytes32 actionId,
         uint256 newDelay,
         address[] memory executors
     ) external returns (uint256 scheduledExecutionId) {
-        require(newDelay <= MAX_DELAY, "DELAY_TOO_LARGE");
         require(isRoot(msg.sender), "SENDER_IS_NOT_ROOT");
 
+        uint256 executionDelay = _getDelayChangeExecutionDelay(_delaysPerActionId[actionId], newDelay);
+
+        bytes32 scheduleDelayActionId = getScheduleDelayActionId(actionId);
+        bytes memory data = abi.encodeWithSelector(this.setDelay.selector, actionId, newDelay);
+        return _scheduleWithDelay(scheduleDelayActionId, address(this), data, executionDelay, executors);
+    }
+
+    /**
+     * @notice Schedules an execution to set the delay for granting permission over `actionId` to `newDelay`.
+     */
+    function scheduleGrantDelayChange(
+        bytes32 actionId,
+        uint256 newDelay,
+        address[] memory executors
+    ) external returns (uint256 scheduledExecutionId) {
+        require(isRoot(msg.sender), "SENDER_IS_NOT_ROOT");
+
+        uint256 executionDelay = _getDelayChangeExecutionDelay(_grantDelays[actionId], newDelay);
+
+        bytes memory data = abi.encodeWithSelector(this.setGrantDelay.selector, actionId, newDelay);
+        // TODO: fix actionId for event (maybe overhaul _scheduleWithDelay?)
+        return _scheduleWithDelay(0x0, address(this), data, executionDelay, executors);
+    }
+
+    function _getDelayChangeExecutionDelay(uint256 currentDelay, uint256 newDelay) private pure returns (uint256) {
+        require(newDelay <= MAX_DELAY, "DELAY_TOO_LARGE");
```

Comment:
We are doing both this check and `isRoot` for `scheduleDelayChange` and `scheduleGrantDelayChange`.

How about grouping the the two of them in a separate helper and calling it at the very beginning of the function?
I find it a bit weird that an internal `_get...` starts with a `require` when this could be solved at the beginning of the external method.